RecoTracker/LSTCore/src/alpaka/Event.dev.cc

```diff
@@ -0,0 +1,1824 @@
+      n_segments_by_layer_endcap_[i] = 0;
+      n_triplets_by_layer_endcap_[i] = 0;
+      n_trackCandidates_by_layer_endcap_[i] = 0;
+      n_quintuplets_by_layer_endcap_[i] = 0;
+    }
+  }
+  if (hitsInGPU) {
+    delete hitsInGPU;
+    delete hitsBuffers;
+    hitsInGPU = nullptr;
+  }
+  if (mdsInGPU) {
+    delete mdsInGPU;
+    delete miniDoubletsBuffers;
+    mdsInGPU = nullptr;
+  }
+  if (rangesInGPU) {
+    delete rangesInGPU;
+    delete rangesBuffers;
+    rangesInGPU = nullptr;
+  }
+  if (segmentsInGPU) {
+    delete segmentsInGPU;
+    delete segmentsBuffers;
+    segmentsInGPU = nullptr;
+  }
+  if (tripletsInGPU) {
+    delete tripletsInGPU;
+    delete tripletsBuffers;
+    tripletsInGPU = nullptr;
+  }
+  if (quintupletsInGPU) {
+    delete quintupletsInGPU;
+    delete quintupletsBuffers;
+    quintupletsInGPU = nullptr;
+  }
+  if (trackCandidatesInGPU) {
+    delete trackCandidatesInGPU;
+    delete trackCandidatesBuffers;
+    trackCandidatesInGPU = nullptr;
+  }
+  if (pixelTripletsInGPU) {
+    delete pixelTripletsInGPU;
+    delete pixelTripletsBuffers;
+    pixelTripletsInGPU = nullptr;
+  }
+  if (pixelQuintupletsInGPU) {
+    delete pixelQuintupletsInGPU;
+    delete pixelQuintupletsBuffers;
+    pixelQuintupletsInGPU = nullptr;
+  }
+
+  if (hitsInCPU != nullptr) {
+    delete hitsInCPU;
+    hitsInCPU = nullptr;
+  }
+  if (rangesInCPU != nullptr) {
+    delete rangesInCPU;
+    rangesInCPU = nullptr;
+  }
+  if (mdsInCPU != nullptr) {
+    delete mdsInCPU;
+    mdsInCPU = nullptr;
+  }
+  if (segmentsInCPU != nullptr) {
+    delete segmentsInCPU;
+    segmentsInCPU = nullptr;
+  }
+  if (tripletsInCPU != nullptr) {
+    delete tripletsInCPU;
+    tripletsInCPU = nullptr;
+  }
+  if (quintupletsInCPU != nullptr) {
+    delete quintupletsInCPU;
+    quintupletsInCPU = nullptr;
+  }
+  if (pixelTripletsInCPU != nullptr) {
+    delete pixelTripletsInCPU;
+    pixelTripletsInCPU = nullptr;
+  }
+  if (pixelQuintupletsInCPU != nullptr) {
+    delete pixelQuintupletsInCPU;
+    pixelQuintupletsInCPU = nullptr;
+  }
+  if (trackCandidatesInCPU != nullptr) {
+    delete trackCandidatesInCPU;
+    trackCandidatesInCPU = nullptr;
+  }
+  if (modulesInCPU != nullptr) {
+    delete modulesInCPU;
+    modulesInCPU = nullptr;
+  }
+}
+
+void SDL::Event<SDL::Acc>::addHitToEvent(std::vector<float> x,
+                                         std::vector<float> y,
+                                         std::vector<float> z,
+                                         std::vector<unsigned int> detId,
+                                         std::vector<unsigned int> idxInNtuple) {
+  // Use the actual number of hits instead of a max.
+  unsigned int nHits = x.size();
+
+  // Initialize space on device/host for next event.
+  if (hitsInGPU == nullptr) {
+    hitsInGPU = new SDL::hits();
+    hitsBuffers = new SDL::hitsBuffer<Dev>(nModules_, nHits, devAcc, queue);
+    hitsInGPU->setData(*hitsBuffers);
```

Comment:
These lines make me more confused of the relation of `hits` and `hitsBuffer` classes (and others that follow the same pattern). Given that `hits` is the base class of `hitsBuffer`, why is a separate `hits` object constructed?

Reply:
- hitsBuffers when created allocate actual storage
- hitsInGPU is a struct of pointers to data (on device) in these buffers